owsi-core/owsi-core-components/owsi-core-component-wicket-more/src/main/java/fr/openwide/core/wicket/more/rendering/service/AbstractRendererServiceImpl.java

```diff
@@ -0,0 +1,169 @@
+package fr.openwide.core.wicket.more.rendering.service;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import org.apache.wicket.Application;
+import org.apache.wicket.Localizer;
+import org.apache.wicket.util.convert.IConverter;
+import org.bindgen.BindingRoot;
+import org.javatuples.Triplet;
+
+import com.google.common.collect.Maps;
+
+import fr.openwide.core.jpa.more.rendering.service.IRendererService;
+import fr.openwide.core.jpa.more.util.fieldpath.model.FieldPath;
+import fr.openwide.core.wicket.more.notification.service.AbstractBackgroundWicketThreadContextBuilder;
+import fr.openwide.core.wicket.more.rendering.EnumRenderer;
+import fr.openwide.core.wicket.more.rendering.Renderer;
+
+public abstract class AbstractRendererServiceImpl extends AbstractBackgroundWicketThreadContextBuilder
+		implements IRendererService, IRendererRegistryService {
+	
+	private Map<Triplet<Class<?>, FieldPath, Class<?>>, Renderer<?>> cache = Maps.newHashMap();
+
+	private Map<Triplet<Class<?>, FieldPath, Class<?>>, Renderer<?>> renderers = Maps.newHashMap();
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public final <T> Renderer<? super T> findRenderer(Class<?> rootType, FieldPath path, Class<T> valueType) {
+		Triplet<Class<?>, FieldPath, Class<?>> parameters = Triplet.<Class<?>, FieldPath, Class<?>>with(rootType, path, valueType);
+		
+		Renderer<? super T> renderer = (Renderer<? super T>) cache.get(parameters);
+		if (renderer != null) {
+			return renderer;
+		} else {
+			renderer = findRendererNoCache(rootType, path, valueType);
+			cache.put(parameters, renderer);
+			return renderer;
+		}
+	}
+	
+	@SuppressWarnings("unchecked")
+	private <T> Renderer<? super T> findRendererNoCache(Class<?> rootType, FieldPath path, Class<T> valueType) {
+		Renderer<? super T> renderer = (Renderer<? super T>) renderers.get(Triplet.with(rootType, path, valueType));
+		if (renderer != null) {
+			return renderer;
+		}
+		renderer = (Renderer<? super T>) renderers.get(Triplet.with(rootType, path, null));
+		if (renderer != null) {
+			return renderer;
+		}
+		renderer = (Renderer<? super T>) renderers.get(Triplet.with(rootType, null, valueType));
+		if (renderer != null) {
+			return renderer;
+		}
+		renderer = (Renderer<? super T>) renderers.get(Triplet.with(null, null, valueType));
+		if (renderer != null) {
+			return renderer;
+		}
+		return findDefaultRenderer(valueType);
+	}
+	
+	protected final <T> Renderer<? super T> findDefaultRenderer(final Class<T> valueType) {
```

Comment:
Should probably be private instead of protected final (or if there is a reason, there should be a comment). Will look into it.